pkg/controller/bundlec/bundle_sync_task.go

```diff
@@ -293,42 +292,40 @@ func (st *bundleSyncTask) updateBundle() error {
 	if err != nil {
 		return errors.Wrap(err, "failed to update bundle")
 	}
-	st.logger.Sugar().Debugf("Set bundle status to %s", &bundleUpdated.Status)
 	st.logger.Sugar().Debugf("Set bundle finalizers to %s", bundleUpdated.Finalizers)
 	return nil
 }
 
+func (st *bundleSyncTask) updateBundleStatus() error {
+	bundleUpdated, err := st.bundleClient.Bundles(st.bundle.Namespace).UpdateStatus(st.bundle)
+	if err != nil {
+		return errors.Wrap(err, "failed to update bundle status")
+	}
+	st.logger.Sugar().Debugf("Set bundle status to %s", &bundleUpdated.Status)
+	return nil
+}
+
 func (st *bundleSyncTask) handleProcessResult(retriable bool, processErr error) (bool /*retriable*/, error) {
 	if processErr != nil && api_errors.IsConflict(errors.Cause(processErr)) {
 		return retriable, processErr
 	}
 	if processErr == context.Canceled || processErr == context.DeadlineExceeded {
 		return false, processErr
 	}
-
-	bundleUpdated := false
-
 	if st.newFinalizers != nil {
 		// Update finalizers
 		st.bundle.Finalizers = st.newFinalizers
-		// Set status to "InProgress: True"
-		// (there will be one more iteration for resource sync that will set an appropriate status)
-		// TODO: Setting status to "InProgress" here might be unnecessary
-		// (if the Bundle spec wasn't updated during the last update, the status should be kept with no changes)
-		// The better solution could be storing "ObservedGeneration" in the status to reflect
-		// whether the status is up-to-date with the spec or stale
-		st.bundle.Status.Conditions = []cond_v1.Condition{
-			{Type: smith_v1.BundleInProgress, Status: cond_v1.ConditionTrue},
-			{Type: smith_v1.BundleReady, Status: cond_v1.ConditionFalse},
-			{Type: smith_v1.BundleError, Status: cond_v1.ConditionFalse},
-		}
-		_, err := st.updateObjectsToDeleteStatus()
+		err := st.updateBundle()
```

Comment:
I'm pretty sure that `status` subresource allows modifying finalizers.
I have recently found this code in Service Catalog https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_instance.go#L2420 and double checked that it actually works.
So unless CRD implementation is different, you shouldn't need to invoke this update separately from status update.
Also, there should only be one update request per processing, as we all learned the hard way.

Reply:
> I'm pretty sure that status subresource allows removing finalizers.

But it should not. See the design doc and the issues I linked above.

> So unless CRD implementation is different, you shouldn't need to invoke this update separately from status update.

When adding finalizers we don't perform bundle processing and hence do not update its status.

> Also, there should only be one update request per processing, as we all learned the hard way.

Yes. That is how it was and this PR does not change this behaviour.